src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,377 @@
+                                      uint8_t              aSeriesId,
+                                      const otLinkMetrics &aLinkMetricsFlags)
+{
+    otError                error;
+    LinkMetricsTypeIdFlags typeIdFlags[kMaxTypeIdFlags];
+    uint8_t                typeIdFlagsCount = GetTypeIdFlagsFromOtLinkMetricsFlags(typeIdFlags, aLinkMetricsFlags);
+
+    error = SendLinkMetricsQuery(aDestination, aSeriesId, typeIdFlags, typeIdFlagsCount);
+
+    return error;
+}
+
+otError LinkMetrics::AppendLinkMetricsReport(Message &aMessage, const Message &aRequestMessage)
+{
+    otError       error = OT_ERROR_NONE;
+    Tlv           tlv;
+    uint8_t       queryId;
+    bool          hasQueryId  = false;
+    uint8_t       length      = 0;
+    uint16_t      startOffset = aMessage.GetLength();
+    uint16_t      offset;
+    uint16_t      endOffset;
+    otLinkMetrics linkMetrics;
+
+    memset(&linkMetrics, 0, sizeof(linkMetrics));
+
+    SuccessOrExit(error = Tlv::FindTlvOffset(aRequestMessage, Mle::Tlv::Type::kLinkMetricsQuery, offset));
+    VerifyOrExit(aRequestMessage.Read(offset, sizeof(tlv), &tlv) == sizeof(tlv), error = OT_ERROR_PARSE);
+
+    offset += sizeof(Tlv);
+    endOffset = offset + tlv.GetLength();
+
+    while (offset < endOffset)
+    {
+        VerifyOrExit(aRequestMessage.Read(offset, sizeof(tlv), &tlv) == sizeof(tlv), error = OT_ERROR_PARSE);
+
+        switch (tlv.GetType())
+        {
+        case kLinkMetricsQueryId:
+            SuccessOrExit(error = Tlv::ReadUint8Tlv(aRequestMessage, offset, queryId));
+            hasQueryId = true;
+            break;
+
+        case kLinkMetricsQueryOptions:
+            for (uint16_t i = offset + sizeof(tlv); i < offset + sizeof(tlv) + tlv.GetLength();
+                 i += sizeof(LinkMetricsTypeIdFlags))
+            {
+                uint8_t typeIdFlags;
+                VerifyOrExit(aRequestMessage.Read(i, sizeof(typeIdFlags), &typeIdFlags) == sizeof(typeIdFlags),
+                             error = OT_ERROR_PARSE);
+                switch (typeIdFlags)
+                {
+                case kTypeIdFlagPdu:
+                    VerifyOrExit(!linkMetrics.mPduCount, error = OT_ERROR_PARSE);
+                    linkMetrics.mPduCount = true;
+                    break;
+
+                case kTypeIdFlagLqi:
+                    VerifyOrExit(!linkMetrics.mLqi, error = OT_ERROR_PARSE);
+                    linkMetrics.mLqi = true;
+                    break;
+
+                case kTypeIdFlagLinkMargin:
+                    VerifyOrExit(!linkMetrics.mLinkMargin, error = OT_ERROR_PARSE);
+                    linkMetrics.mLinkMargin = true;
+                    break;
+
+                case kTypeIdFlagRssi:
+                    VerifyOrExit(!linkMetrics.mRssi, error = OT_ERROR_PARSE);
+                    linkMetrics.mRssi = true;
+                    break;
+
+                default:
+                    break;
```

Comment:
The code does not seem to handle the `E` flag case. As I mentioned from https://github.com/openthread/openthread/pull/5481#discussion_r493221529, if we see the `E` flag set we need to not only skip over the Type ID byte but also the byte that follows it.

Thread spec defines the behavior of "Type ID Flags" with `E` flag:
```1: An additional second flag byte after the current 1-byte flags```

Even though the use of `E` flag is reserved for future, we should parse such a case correctly and skip over two bytes (otherwise we can incorrectly handle the second byte as a new type id flag).